### ISSUE TITLE
Field equals throws ClassCastException

### ISSUE THREAD

**Issue by vyletpet** [jOOQ/jOOQ#8435]
### Expected behavior and actual behavior:
I would expect equals method to return false when comparing different Class but it actually end up with class cast exception when we use our implementation of Field interface or use Mock.

### Steps to reproduce the problem 

```
class FieldEqualsTest extends Specification {

    def "Failing Field equals"() {
        when:
        def field = DSL.field("someField")
        def myField = Mock(Field)

        then:
        field != myField
    }
}
```

The stacktrace is 

```
Caused by: java.lang.ClassCastException: com.sun.proxy.$Proxy16 cannot be cast to org.jooq.QueryPartInternal
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:399)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.DefaultDSLContext.renderInlined(DefaultDSLContext.java:702)
	at org.jooq.impl.AbstractQueryPart.equals(AbstractQueryPart.java:150)
	at org.jooq.impl.AbstractNamed.equals(AbstractNamed.java:109)
```

There is check for QueryPart instance but then it is typed to QueryPartInternal which is a different interface.

### Versions:

- jOOQ: 3.11.9-10

Used to work on 3.10.5

**Comment by lukaseder** [comment]
Thanks for your report. For the record, what's `Mock`?

One general rule of the jOOQ API is that you don't implement the jOOQ API, only jOOQ's internals implement it according to some internal, undocumented rules. E.g. jOOQ internals can safely assume that every `QueryPart` is also a `QueryPartInternal`.

Of course, you could tamper with these internals and mock them, in case of which you have to abide to those undocumented rules. SO, the way I see this, I'd say things work as designed (and you actually shouldn't mock `Field`).

Now, before we continue discussing the above perceived regression, would you mind to explain what exactly you were trying to achieve with this mock, and possibly others?

**Comment by vyletpet** [comment]
Mock is simple mocking mechanism from Spock Framework http://spockframework.org/spock/docs/1.3/interaction_based_testing.html#_creating_mock_objects

We used Mock(Field) to simplify our unit tests.

We have already migrated to latest version and updated our tests so this issue is not blocking us in anyway.

We just noticed this regression while fixing our test and wanted to let you know and see if it was intended change or not. 

The fix is fairly easy but it's up you to decide if it is worth fixing as I can see the gain here is very little too.

**Comment by lukaseder** [comment]
Thanks for the additional feedback. I'm not sure what you mean by this:

> The fix is fairly easy but it's up you to decide if it is worth fixing as I can see the gain here is very little too.

I don't think that a thorough fix is easy here. I mean, "fixing" in this context means making sure that such an exception can *never* happen when people mock jOOQ API that has not been designed for being implemented or mocked by users. In a way, we would make new guarantees that we would have to maintain in order to be sure not to run into such "regressions" again.

I don't think this is worth it, at least not until I understand all the involved use cases very well. Currently, I don't see the use case of mocking jOOQ API...

So, I'll close this as won't fix.

Thanks anyway for taking the time to report this!
